### PR TITLE
RDRP-1171 new functions for filtering

### PR DESCRIPTION
**Pull request by AnneONS**
## Pull Request submission

The create_mask function in imputation_helpers.py previously had limited use as it doesn't work if not all the columns needed for all possible checks are in the dataframe.

For example, if a unit test doesn't include col "604", it's not possible to use create_mask at all.

A check that the correct column is in the dataframe is needed, and also, a way of calling each of the masks individually so we no longer need all possible

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/utils/filter_helpers.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
+import pandas as pd
+from typing import List
+
+
+def create_notnull_mask(df: pd.DataFrame, col: str) -> pd.Series:
+    """Return a mask for string values in column col that are not null."""
+    return df[col].str.len() > 0
```
Comment (adriano-lopresti) [inline src/utils/filter_helpers.py:7]: With null here you mean that it's read in as an empty string, right?
I'm not sure what will happen if the column you are reading is a numeric one, and you have a null value.. 
maybe converting it into string will give you 'nan' or '<na>' the length of which is not zero?

Not saying is wrong, but worth being careful

File: src/utils/filter_helpers.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    msg = ""
+    for col in cols:
+        if col not in df.columns:
+            msg += f"Column {col} does not exist in this dataframe. "
+    if msg != "":
+        raise ValueError(msg)
+
+
+def get_clear_status_mask(df: pd.DataFrame) -> pd.Series:
+    check_cols_in_df(df, ["status"])
```
Comment (adriano-lopresti) [inline src/utils/filter_helpers.py:29]: General comment: if the column exists, this returns an error.
Didn't we said that you wanted also to have the possibility of applying the filter but with the option of avoiding the error of the column does not exist? in this case, whouldn't it make sense to add another boolean argument called, say, "error_if_no_column", which, if a column doesn't exist, makes the "check_cols_in_df" skip the filter rather than throwing an error?

Or did you want to avoid an error if even a column on which you were not filtering, didn't exist? In this case, I think you are implementing it correctly.

File: src/utils/filter_helpers.py (hunk 1 above)
Comment (adriano-lopresti) [reply src/utils/filter_helpers.py:7]: PS: I see from the unit test that it seems to work, so I am ok with it..
But my comment still stands as something to be careful in case all seems to look fine in the unit tests but then when real "edge-case" data are used then it suddenly doesn't behave as expected

File: tests/test_utils/test_filter_helpers.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        result_mask = create_mask(df, options)
+        assert_series_equal(result_mask, expected_mask, check_names=False)
+
+    def test_not_mor_imputed_longform(self, create_input_df):
+        df = create_input_df
+        options = ["not_mor_imputed", "longform_only"]
+        expected_mask = pd.Series([False, False, True, True, True, False])
+        result_mask = create_mask(df, options)
+        assert_series_equal(result_mask, expected_mask, check_names=False)
+
```
Comment (adriano-lopresti) [inline tests/test_utils/test_filter_helpers.py:199]: I'm happy with the tests so far.
You still seem to be working to the test below.

**Review by adriano-lopresti — COMMENTED** [review]
all good. I just have a couple of comments

File: src/utils/filter_helpers.py (hunk 1 above)
Comment (AnneONS) [reply src/utils/filter_helpers.py:7]: This function was introduced by GZ who was looking at all the different ways nulls work. Maybe we should change its name so we know to only use it for string cols?

File: src/utils/filter_helpers.py (hunk 2 above)
Comment (AnneONS) [reply src/utils/filter_helpers.py:29]: it's the latter, and the way the other functions are set up will allow us to do everything we want to.
I think we really do want to throw an error if the code tells us to filter on a column and it doesn't exist.

**Review by AnneONS — COMMENTED** [review]
Here are my responses to your questions @adriano-lopresti  
thanks for the review.
